### ISSUE TITLE
CSD windows under Compiz do not reflect visual changes related to widget's state.

### ISSUE THREAD

**Issue by LarkenVanDort** [compiz-reloaded/compiz#71]
When using Compiz, the CSD windows seem to never change the state from default to backdrop or vice versa.
In other words, if given CSD window gains or loses focus, there is no visual change.

Below pictures should capture this problem quite clearly.

This is how it **should** look like (and how it looks like on Compiz 0.9x)
![proper_behavior](https://cloud.githubusercontent.com/assets/28027283/25412933/509624a0-2a1e-11e7-9713-4f3a38c7eb77.png)

This is how it looks like on Compiz Reloaded (both windows seem active, but in reality, only the top one has focus)
![both_pseudo_active](https://cloud.githubusercontent.com/assets/28027283/25412944/65011a9e-2a1e-11e7-8a1c-cf04ed6d79c0.png)

Below can also happen - under rare circumstances - for example when replacing xfwm4 with Compiz, when CSD apps are already running (it might not be clear, but in this picture, again, the top window has focus)
![both_pseudo_backdrop](https://cloud.githubusercontent.com/assets/28027283/25412946/66be7d36-2a1e-11e7-8437-cf920962e796.png)

In the last scenario, it is possible to force given window to change its visuals from backdrop to default, by - for example - maximizing it. Afterwards it will stay (pseudo)active no matter what.

I have tested number of themes with no luck (and all of them work properly on Compiz 0.9x).
Is there any known workaround?

**Comment by soreau** [comment]
This might be related to advertising _GTK_FRAME_EXTENTS, for which there is a [pull request](https://github.com/compiz-reloaded/compiz/pull/8). I think there are some issues to work out still but I'm not sure as I cannot reproduce the problems described there because I can't find an application that exhibits the bug.

**Comment by soreau** [comment]
I just tested that code and it doesn't seem to make a difference. xfwm4 seems to work properly whereas compiz doesn't do whatever is needed to make it work right.

**Comment by LarkenVanDort** [comment]
Thank you for your reply soreau.

I have to revise my former statement. This bug is not limited to CSD windows. It happens on all gtk3 windows. I believe this issue is related to **_NET_WM_STATE_FOCUSED** flag
(https://bugs.launchpad.net/compiz-core/+bug/904806).

**Comment by raveit65** [comment]
The backdrop state is also broken with marco itself, and the gtk-decorator is build from marco.
As far i know compiz-0.9 is build with metacity where this implementation works.

**Comment by lukefromdc** [comment]
Confirm this is broken in both marco and compiz-0.8

**Comment by lukefromdc** [comment]
I managed a rough port of https://bazaar.launchpad.net/~compiz-team/compiz/0.9.13/revision/2919
to compiz-reloaded, but still have some bugs to deal with

https://github.com/lukefromdc/compiz/tree/dev-focus_WIP

At the moment I sometimes get a crash, these warnings will show up in ~/xsettings-errors, not sure if they are related:
```
Window manager warning: Buggy client sent a _NET_ACTIVE_WINDOW message with a timestamp of 0 for 0x3400086 (Untitled D)
Window manager warning: meta_window_activate called by a pager with a 0 timestamp; the pager needs to be fixed.
```

**Comment by soreau** [comment]
@lukefromdc Nice work, and thanks for making a branch specific to this bug. When you say there are some bugs to deal with, are you referring to the crashes you're getting or functionality of the patch? If you're using X 1.19.6, it might be the crash others are experiencing. You should be able to get a backtrace with the crash handler plugin.

Regarding the patch, I left some in-line comments.

**Comment by lukefromdc** [comment]
Rebuilding plugins made the plugins themselves work properly but the crashes have been a bear, I do NOT get them with my extents branch by itself.  Might be a race condition somewhere, and I really don't understand how the code to reset the window state actually works. Big problem has been adding or removing _NET_WM_STATE_FOCUSED without affecting anything else, plus figuring out the right way to toggle it. The C++ code from the 0.9 branch is so different in how its laid out

**Comment by lukefromdc** [comment]
Also, the crashes most commonly occurred on opening a CSD window or on compiz startup-and if compiz kept running, sometimes the code to set :backdrop via setting _NET_WM_STATE_FOCUSED  worked and sometimes it did not. In my initial tests I had a ton of windows open and it seemed much more stable, and I was able to freely focus and defocus a gnome-disks window, and transfer focus between it and a gedit window. I've seen it work, but what I've got so far has a lot of issues. Still, this is the first time I've ever seen CSD decorations respond to focus and :backdrop set with a build of compiz 0.8 branch so I must be getting somewhere

**Comment by soreau** [comment]
Are you able to get a backtrace with the crash handler plugin?

**Comment by soreau** [comment]
FWIW, I was able to get it working with [this](https://www.hastebin.com/raw/ivojapibej) diff.

**Comment by muktupavels** [comment]
While you work on this, I suggest to do also this:
https://git.xfce.org/xfce/xfwm4/commit/?id=0feb29e78bb3b36387950dc2fb8d991ac875a374

**Comment by soreau** [comment]
@muktupavels Thanks for pointing this out. [Here](https://www.hastebin.com/raw/dogekukupu)'s a v2 that should do it.

**Comment by muktupavels** [comment]
@soreau it does not seem to be complete... Or in other words, it seems that you never set appears focused on docks or splashscreens. Looking at diff seems that it only tries to not remove that state.

**Comment by soreau** [comment]
@muktupavels The line `changeWindowState (w, w->state | CompWindowStateFocusedMask);` is actually what sets it when FocusIn event is received when the window is first created. I'm not sure FocusIn is guaranteed on all windows though so it might be better to enable the state in MapNotify or CreateNotify for 'appear focused' windows.

**Comment by muktupavels** [comment]
Hmm, what if compiz is started after window has created and/or mapped? In `metacity` there are `set_net_wm_state` function that are called for multiple places... I will try to test how this work in 0.9.x.

**Comment by soreau** [comment]
That's true, create or map events might not be the best place. If compiz is started after the window was created, addWindow() is still called for each window so maybe it could be set there, after type is assigned (which is when recalcWindowType() is called).

**Comment by lukefromdc** [comment]
Those patches work beautifully on my system, both with and without using the _GTL_FRAME_EXTENTS patch with them. Plugins were rebuilt for both tests and worked fine.  Will be using this daily until more patches come in, probably should submit a PR for https://github.com/lukefromdc/compiz/tree/dev-focus_WIP

**Comment by muktupavels** [comment]
This is what I did for compiz 0.9.x:
https://code.launchpad.net/~muktupavels/compiz/make-docks-appear-focused/+merge/339421

**Comment by lukefromdc** [comment]
Restarting compiz has in fact created issues with already mapped windows, trying to find a fix now

**Comment by lukefromdc** [comment]
Issues are:

1: existing windows sometimes not mapped after a compiz start/restart especially from fusion-icon (uses --ignore-desktop-hints), and especially if a CSD window or any window running as root (usually Nemo) was showing before start.

2: On rare occasions after a compiz restart. one or more windows will refuse to acccept focus until other windows have been clicked on (raising and focussing them). Again worst if a CSD window is showing. These tests are with my combined branch.

In compiz-reloaded from master or just advertising GtkFrameExtent support (my original branch), there was already a problem with mate-panel not being mapped after  a compiz restart from fusion-icon until the desktop was clicked on. so this may relate to a pre-existing issue

**Comment by muktupavels** [comment]
@lukefromdc how last comment are related to this bug? sounds like separate problems/bugs...

**Comment by lukefromdc** [comment]
The two new focus/compiz restart bugs mentioned above seem to be gone after rebasing my branch(es) after https://github.com/compiz-reloaded/compiz/commit/9752b0cbcbe9a28a34a67f3142f5cb7b9d33e4f6 
Since that commit could not have affected this, some underlying change must have helped. Now I have compiz restarting just fine, even with CSD windows showing, and with the focus work in my branch the window states of CSD windows reflect focus/backdrop just fine. I now seem to have that working without any adverse side effects, so submitting it as a PR

**Comment by lukefromdc** [comment]
https://github.com/compiz-reloaded/compiz/pull/117
